File: src/Site/arena.rs
use std::borrow::Borrow;

use html_escape::decode_html_entities;
use regex::Regex;

use crate::event::{Event, Location};
use crate::site::eventim::Eventim;
use crate::site::{Filter, Site};
use crate::tools::date::parse_short_date;
use crate::tools::HTTP;

const URL: &str = "https://www.quarterback-immobilien-arena.de";

pub(crate) struct Arena {
    location: Location,
    location_id: i8,
    eventim_id: String,
}

impl Arena {
    pub(crate) fn new_red_bull() -> Self {
        Self {
            location: Location {
                slug: "rb".to_string(),
                name: "Red Bull Arena".to_string(),
                website: URL.to_owned() + "/red-bull-arena",
            },
            location_id: 2,
            eventim_id: "red-bull-arena-16304".to_string(),
        }
    }

    pub(crate) fn new_quarterback() -> Self {
        Self {
            location: Location {
                slug: "qi".to_string(),
                name: "QUARTERBACK Immobilien ARENA".to_string(),
                website: URL.to_owned() + "/quarterback-immobilien-arena",
            },
            location_id: 1,
            eventim_id: "quarterback-immobilien-arena-leipzig-383".to_string(),
        }
    }

    pub(crate) fn new_festwiese() -> Self {
        Self {
            location: Location {
                slug: "fw".to_string(),
                name: "Festwiese Leipzig".to_string(),
                website: URL.to_owned() + "/festwiese-leipzig",
            },
            location_id: 3,
            eventim_id: "festwiese-leipzig-7410".to_string(),
        }
    }
}

impl Site for Arena {
    fn get_locations(&self) -> Vec<Location> {
        return vec![self.location.clone()];
    }

    fn fetch_events(&self) -> Vec<Event> {
        let http = HTTP::new();
        let mut result = Vec::new();

        let eventim = Eventim::new(self.eventim_id.to_string(), http.borrow());
        let reg: Regex = Regex::new(
            "(?si)<div class=\"event\".*?<a href=\"(.*?)\">.*?<source srcset=\"(.*?)\" media=\"[(]max-width: 320px[)]\">.*?<div>\\w+,\\s+(\\d\\d\\.\\d\\d\\.\\d\\d\\d\\d)</div>.*?<h2>(.*?)</h2>",
        )
        .unwrap();

        for page in 1..100 {
            let html = http.get(&(URL.to_owned() + "/events-tickets/events?tx_ifabeventmanagementextend_searchfilter[eventSearch][category]=1"
            +"&tx_ifabeventmanagementextend_searchfilter[eventSearch][location]="+ &self.location_id.to_string()
            +"&tx_ifabeventmanagement_events[@widget_0][currentPage]="+&page.to_string())
            );

            for captures in reg.captures_iter(html.as_str()) {
                let evt = Event::new(
                    decode_html_entities(&captures[4]).to_string(),
                    parse_short_date(&captures[3]).and_hms(0, 0, 0),
                    self.location.borrow(),
                    URL.to_owned() + captures[1].borrow(),
                    Option::Some(URL.to_owned() + captures[2].borrow()),
                );
                if eventim.is_it_metal(evt.borrow()) {
                    result.push(evt);
                }
            }
            if !html.contains("<li class=\"next\">") {
                break;
            }
        }

        result
    }
}

File: src/Site/Arena.php
<?php

declare(strict_types=1);

namespace MetalLE\Site;

use MetalLE\Event\Event;
use MetalLE\Event\Location;
use MetalLE\Site\Shop\Eventim;
use Traversable;

class Arena implements Site
{
    private const URL = 'https://www.quarterback-immobilien-arena.de';

    public function __construct(
        private Crawler $http = new Crawler(),
    ) {
    }

    public function getIterator(): Traversable
    {
        $rb = new Location('rb', 'Red Bull Arena', self::URL . '/red-bull-arena');
        yield from $this->crawlContent(2, $rb, new Eventim('red-bull-arena-16304'));

        $immo = new Location('qi', 'QUARTERBACK Immobilien ARENA', self::URL . '/quarterback-immobilien-arena');
        yield from $this->crawlContent(1, $immo, new Eventim('quarterback-immobilien-arena-leipzig-383'));

        $fw = new Location('fw', 'Festwiese Leipzig', self::URL . '/festwiese-leipzig');
        yield from $this->crawlContent(3, $fw, new Eventim('festwiese-leipzig-7410'));
    }

    private function crawlContent(int $locationId, Location $location, Eventim $eventim, int $page = 1): \Generator
    {
        $html = $this->http->get(
            self::URL . '/events-tickets/events?tx_ifabeventmanagementextend_searchfilter[eventSearch][category]=1'
            . '&tx_ifabeventmanagementextend_searchfilter[eventSearch][location]=' . $locationId
            . '&tx_ifabeventmanagement_events[@widget_0][currentPage]=' . $page
        );

        if (preg_match_all(
            '#<div class="event".*<a href="(.*)">.*<source srcset="(.*)" media="\(max-width: 320px\)">.*<div>\w+,\s+(\d\d\.\d\d\.\d\d\d\d)</div>.*<h2>(.*)</h2>#isU',
            $html,
            $matches,
            PREG_SET_ORDER
        )) {
            foreach ($eventim->fetchDates() as $date) {
                foreach ($matches as $match) {
                    if ($date->format('d.m.Y') === $match[3]) {
                        yield new Event(html_entity_decode($match[4]), $date, $location, self::URL . $match[1], self::URL . $match[2]);
                        break;
                    }
                }
            }
        }

        if (str_contains($html, '<li class="next">')) {
            yield from $this->crawlContent($locationId, $location, $eventim, $page + 1);
        }
    }
}